sys/linux/io_uring.txt

```diff
@@ -52,7 +117,529 @@ io_sqring_offsets {
 	resv2		const[0, int64]
 }
 
-io_uring_files_update {
-	offset	int32
-	fds	ptr[in, array[fd]]
+io_cqring_offsets {
+	head		const[0, int32]
+	tail		const[0, int32]
+	ring_mask	const[0, int32]
+	ring_entries	const[0, int32]
+	overflow	const[0, int32]
+	cqes		const[0, int32]
+	flags		const[0, int32]
+	resv1		const[0, int32]
+	resv2		const[0, int64]
+}
+
+# NEED: To avoid so much boilerplate, use call templates (when implemented).
+syz_io_uring_submit$IORING_OP_NOP(sq_ring_ptr sq_ring_ptr, sqes_ptr sqes_ptr, sqe ptr[in, io_uring_sqe_nop], sqes_index int32)
```

Comment:
We may try to reduce duplication by accepting a union with all of the ops:
```
syz_io_uring_submit(sq_ring_ptr sq_ring_ptr, sqes_ptr sqes_ptr, sqe ptr[in, io_uring_sqe_u], sqes_index int32)

io_uring_sqe_u [
    IORING_OP_NOP    io_uring_sqe_nop
    IORING_OP_FSYNC  io_uring_sqe_fsync
    ...                    
] [varlen]
```

and additionally maybe a template for _all_ operations b/c lots of fields are the same across all ops:
```
type io_uring_sqe_t[OP, FD, OFF_ADDR] {
	opcode		const[OP, int8]
	flags		flags[iosqe_flags, int8]
	ioprio		const[0, int16]
	fd		FD
	off_addr	OFF_ADDR
        ...
}
```

ops that use off_addr union will pass, say, OFF_ADDR = fileoff[int64], while others will pass const[0, int64] for OFF_ADDR.

Though, the downside of this may be (as with any template) that descriptions will become more cryptic.

Reply:
Sure, it should be cleaner this way. 

I am not sure how many fields we can share using a union for io_uring_sqe though -- the usage of fields substantially change among different operations. It should still be useful at least for `opcode` and `flags`, and cleaner with less repetitions of the field names.

I am going to delay doing this for a while and get it done once the other TODOs are done.

Reply:
[Refactored](https://github.com/google/syzkaller/pull/1926/commits/afe1176eb3a8f28490c2d61ac2510ba14826ebf4)